### PR TITLE
ws: Look for "login-data" in spawn login results.

### PR DESCRIPTION
**Pull request by petervo**
Allows custom login commands to add data to the HTTP response on a successful login.

### CODE REVIEW

File: src/ws/cockpitcreds.c [hunk 1]
```diff
@@ -219,6 +224,43 @@ cockpit_creds_get_csrf_token (CockpitCreds *creds)
...
+ * @object: A JsonObject to store
+ *
+ * It is an error to call this function
+ * if login data has already been set.
+ *
+ */
+
+void
+cockpit_creds_set_login_data (CockpitCreds *creds,
+                              JsonObject *object)
```
Comment (stefwalter) [inline src/ws/cockpitcreds.c]: The idea is that CockpitCreds is an immutable class ... with the exception of the atomic cockpit_creds_poison() stuff, which just sets a flag.

File: src/ws/cockpitauth.c [hunk 2]
```diff
@@ -550,8 +551,16 @@ create_creds_for_spawn_authenticated (CockpitAuth *self,
...
 
-  g_free (csrf_token);
+  if (!cockpit_json_get_object (results, "login-data", NULL, &login_data))
+    {
+      g_warning ("received bad login-data from %s", sl->command);
+      login_data = NULL;
+    }
 
+  if (login_data)
+    cockpit_creds_set_login_data (creds, login_data);
```
Comment (stefwalter) [inline src/ws/cockpitauth.c]: An alternative would be to return the JsonObject from cockpit_auth_login_finish() and pass it to send_login_response(). We don't need to keep this data loaded in cockpit-ws, and thus it doesn't really need to be part of CockpitCreds.

Comment (petervo) [inline src/ws/cockpitauth.c]: The reason i did it this way is that if you call /login with a valid cookie that it has a session for it just responds straight away so if we want to include login-data in that response we need to keep it around.

Comment (stefwalter) [inline src/ws/cockpitauth.c]: Makes sense. Lets just do it in an immutable way.

**Comment by petervo** [comment]
Updated to keep cockpitcreds immutable. 

Re-parsing the json string data is pretty annoying but seemed less annoying than implementing a json object clone function since just taking a ref means that the object can still be changed by someone else, the _dup functions in json lib seem to just increment refs for objects. It also has the minor advantage that it doesn't require changing the the cockpit_creds_new function declaration. But let me know if you think a different approach would be better.

File: src/common/cockpitjson.h [hunk 3]
```diff
@@ -81,6 +81,11 @@ gboolean       cockpit_json_get_array         (JsonObject *object,
                                                JsonArray *defawlt,
                                                JsonArray **value);
 
+gboolean       cockpit_json_get_object        (JsonObject *options,
+                                               const gchar *name,
+                                               JsonObject *defawlt,
```
Comment (dperpeet) [inline src/common/cockpitjson.h]: this should be declared `const` too

File: src/common/test-json.c [hunk 4]
```diff
@@ -238,6 +239,39 @@ test_get_array (TestCase *tc,
...
+  ret = cockpit_json_get_object (tc->root, "unknown", defawlt, &value);
+  g_assert (ret == TRUE);
+  g_assert (value == defawlt);
+
+  ret = cockpit_json_get_object (tc->root, "number", NULL, &value);
+  g_assert (ret == FALSE);
+
+  ret = cockpit_json_get_object (tc->root, "array", NULL, NULL);
+  g_assert (ret == FALSE);
+
```
Comment (dperpeet) [inline src/common/test-json.c]: For the sake of completeness we should add a test here for the case "query existing entry, pass default value, result is not the default value."
This shouldn't be an issue, but would improve the coverage.

File: src/ws/cockpitauth.c [hunk 5]
```diff
@@ -511,7 +511,8 @@ static CockpitCreds *
 create_creds_for_spawn_authenticated (CockpitAuth *self,
                                       const char *user,
                                       SpawnLoginData *sl,
-                                      JsonObject *results)
+                                      JsonObject *results,
+                                      const char *raw_data)
```
Comment (dperpeet) [inline src/ws/cockpitauth.c]: Why are `user` and `raw_data` in this function of type `const char *` and not `const gchar *`? The values passed to them are `gchar`.

**Comment by dperpeet** [comment]
Minor issues, otherwise this looks good.

File: src/common/test-json.c (hunk 4 above)
Comment (petervo) [inline src/common/test-json.c]: Not really there is no special code path for that, we do check that the default NULL, is not returned i think that covers the use case.

File: src/ws/cockpitauth.c (hunk 5 above)
Comment (petervo) [inline src/ws/cockpitauth.c]: Changed it, but now someone else will complain that I changed old code unnecessarily.

**Pull request closed without merging**